app/frontend/wallet/address-manager.ts

```diff
@@ -3,79 +3,91 @@ import {toBip32StringPath} from './helpers/bip32'
   gapLimit,
+  defaultAddressCount,
   cryptoProvider,
   disableCaching, // good for tests
   isChange,
   blockchainExplorer,
 }) => {
+  // for scheme.v1 we used to derive first defaultAddressCount addresses, make sure we can re-discover them now
+  if (defaultAddressCount > gapLimit) {
+    throw NamedError('ParamsValidationError', 'Invalid default address count')
+  }
+
+  return _AddressManager({
+    addrGen: ByronAddressGenerator(cryptoProvider, accountIndex, isChange),
+    gapLimit,
+    disableCaching, // good for tests
+    blockchainExplorer,
+  })
+}
+const _AddressManager = ({
+  addrGen,
+  gapLimit,
+  disableCaching, // good for tests
+  blockchainExplorer,
+}) => {
+  if (!gapLimit) {
+    throw NamedError('ParamsValidationError', `Invalid gap limit: ${gapLimit}`)
+  }
+
   const deriveAddressMemo = {}
-  const derivationScheme = cryptoProvider.getDerivationScheme()
 
-  validateParams()
+  async function cachedDeriveAddress(index: number) {
+    const memoKey = index
 
-  function validateParams() {
-    if (!gapLimit) {
-      throw NamedError('ParamsValidationError', `Invalid gap limit: ${gapLimit}`)
-    }
-    if (!defaultAddressCount) {
-      throw NamedError(
-        'ParamsValidationError',
-        `Invalid defaultAddressCount: ${defaultAddressCount}`
-      )
+    if (!deriveAddressMemo[memoKey] || disableCaching) {
+      deriveAddressMemo[memoKey] = await addrGen(index)
     }
-  }
 
-  async function discoverAddresses() {
-    switch (derivationScheme.type) {
-      case 'v1':
-        return await discoverAddressesV1()
-      case 'v2':
-        return await discoverAddressesV2()
-      default:
-        throw NamedError(
-          'DerivationSchemeError',
-          `Unexpected derivation scheme: ${derivationScheme.type}`
-        )
-    }
+    return deriveAddressMemo[memoKey].address
   }
 
-  function discoverAddressesV1() {
-    const childIndexBegin = derivationScheme.startAddressIndex
-    const childIndexEnd = childIndexBegin + defaultAddressCount
-    return deriveAddressesBlock(childIndexBegin, childIndexEnd)
+  async function deriveAddressesBlock(beginIndex: number, endIndex: number) {
+    return Promise.all(range(beginIndex, endIndex).map(cachedDeriveAddress))
   }
 
-  async function discoverAddressesV2() {
+  async function discoverAddresses() {
     let addresses = []
-    let childIndexBegin = derivationScheme.startAddressIndex
+    let from = 0
     let isGapBlock = false
 
     while (!isGapBlock) {
-      const currentAddressBlock = await deriveAddressesBlock(
-        childIndexBegin,
-        childIndexBegin + gapLimit
-      )
+      const currentAddressBlock = await deriveAddressesBlock(from, from + gapLimit)
+
       isGapBlock = !(await blockchainExplorer.isSomeAddressUsed(currentAddressBlock))
+
       addresses =
         isGapBlock && addresses.length > 0 ? addresses : addresses.concat(currentAddressBlock)
-      childIndexBegin += gapLimit
+      from += gapLimit
```

Comment:
Previously, we distinguished between `v1` and `v2` address types and we were checking whether "some addresses are used" only for `v2` addresses. What is the reason for not distinguishing anymore and checking for used address for `v1` addresses as well? just to clarify..

Reply:
Assuming `defaultAddressCount < gapLimit` the new code will anyway discover all old addresses. It might discover more of them but that is not a problem for the UI or the rest of functionality ...